src/mw-i18n/services/mw_i18n_test.js

```diff
@@ -269,5 +273,125 @@ describe('mwUi i18n Service', function () {
       expect(i18n.localize(i18nObject)).toEqual('SE:A');
     });
 
+    describe('extend translations', function () {
+      beforeEach(function () {
+        i18nProvider.addResource('i18n/c');
+      });
+
+      it('extends translations with new ones when ressources are already loaded', function (done) {
+        i18n.setLocale('de_DE').then(function () {
+          i18n.extendForLocale('de_DE', {c3: 'DE:C3'});
+
+          expect(i18n.get('c3')).toBe('DE:C3');
+          done();
+        });
+        $rootScope.$digest();
+      });
+
+      it('extends translations with new ones when ressources are not loaded yet', function (done) {
+        i18n.extendForLocale('de_DE', {c3: 'DE:C3'});
+
+        i18n.setLocale('de_DE').then(function () {
+
+          expect(i18n.get('c3')).toBe('DE:C3');
+          done();
+        });
+        $rootScope.$digest();
+      });
+
+      it('replaces translations when ressources are already loaded', function (done) {
+        i18n.setLocale('de_DE').then(function () {
+          i18n.extendForLocale('de_DE', {c2: {c22: {c221: 'DE:REPLACED'}}});
+
+          expect(i18n.get('c2.c22.c221')).toBe('DE:REPLACED');
+          done();
+        });
+        $rootScope.$digest();
+      });
+
+      it('replaces translations when ressources are not loaded yet', function (done) {
+        i18n.extendForLocale('de_DE', {c2: {c22: {c221: 'DE:REPLACED'}}});
+
+        i18n.setLocale('de_DE').then(function () {
+
+          expect(i18n.get('c2.c22.c221')).toBe('DE:REPLACED');
+          done();
+        });
+        $rootScope.$digest();
+      });
+
+      it('keeps replaced translations when changing the locale', function (done) {
+        i18n.extendForLocale('de_DE', {c2: {c22: {c221: 'DE:REPLACED'}}});
+        i18n.extendForLocale('en_US', {c2: {c22: {c221: 'EN:REPLACED'}}});
+        i18n.setLocale('en_US').then(function () {
+
+          i18n.setLocale('de_DE').then(function () {
+            expect(i18n.get('c2.c22.c221')).toBe('DE:REPLACED');
+            done();
+          });
+        });
+        $rootScope.$digest();
+      });
+
+      it('extends translations for multiple locales', function (done) {
+        i18n.extend({
+          'de_DE': {c2: {c22: {c221: 'DE:REPLACED'}}},
+          'en_US': {c2: {c22: {c221: 'EN:REPLACED'}}}
+        });
+
+        i18n.setLocale('en_US').then(function () {
+          expect(i18n.get('c2.c22.c221')).toBe('EN:REPLACED');
+
+          i18n.setLocale('de_DE').then(function () {
+
+            expect(i18n.get('c2.c22.c221')).toBe('DE:REPLACED');
+            done();
+          });
+        });
+        $rootScope.$digest();
+      });
+
+      describe('error handling', function () {
+        it('throws error when locale is not available or invalid', function () {
```

Comment:
could be 2 testcases, one for not available one for invalid (have tests fail for a single reason not for two)